### ISSUE TITLE
Remember Current Disk When Loading a State For Multi-Disk Game

### ISSUE THREAD

**Issue by Zombeaver** [libretro/vice-libretro#293]
I'm not entirely sure if this is possible, but if so I think it would be worth considering.

It's not exactly uncommon to have games that boot on one disk and then switch to another before starting the game, and you may or may not need to switch back to the first one later; and even if it doesn't it may still need to load from the second disk again later on. It's handy to save a state after this initial shuffle has taken place and setting it up to autoload on start. The issue is, it loads the disks in the order that they're listed in the M3U (and loads with the first one listed regardless of whether or not that's actually what's inserted at the time of the state save). This means that in these scenarios I usually end up starting the game, swapping, saving, exiting, changing it to an .auto state, and then flipping the disk order around in the M3U so that it's Disk2.d64 > Disk1.d64, this way it still has the correct disk inserted when loading the state. This means, however, that if you were to reset the game or try to start it without the state it wouldn't work, and it means that "Next Disk" changes to disk 1 which just comes across as awkward.

Preferably the state would remember which disk in the .m3u list is currently inserted and start in that position on load so that none of this is necessary.

**Comment by sonninnos** [comment]
Hmm, the state internally already does that with d64 disks in my tests.

I have `empty.m3u` which has 2 disks labeled differently. I load the m3u, change disk, take a dir list, save state, quit, load state, take a dir list and the listing says the second disk is in.

Same thing with Maniac Mansion. Saved state after disk change, quit, load state, and the correct disk is still inserted, since the next screen loading works.

The problem is that only the frontend shows the wrong disk, because the state changes it behind the curtain. But I guess that could be made to sync up properly too, as long as the savestate file saves the filename too (it actually saves the whole disk data). And since it is already done in the PUAE core, I'll add it to the TODO.

Luckily more and more games are available in single file formats (d81 and carts), so this problem gets rarer and rarer.

**Comment by Zombeaver** [comment]
> The problem is that only the frontend shows the wrong disk, because the state changes it behind the curtain.

Ah _ha_! Interesting, I didn't realize this. Yes, I was going based on the labeling shown in the UI.

> But I guess that could be made to sync up properly too, as long as the savestate file saves the filename too (it actually saves the whole disk data). And since it is already done in the PUAE core, I'll add it to the TODO.

That would be great!

> Luckily more and more games are available in single file formats (d81 and carts), so this problem gets rarer and rarer.

Yep, that's certainly true, and I opt for those wherever possible. Some of the more obscure stuff that isn't an option for though (and probably won't ever be). In some cases I've been able to just convert multiple discs to d81 myself via Dirmaster, but it's not always as simple as that.

**Comment by sonninnos** [comment]
Bad news, the filename is not included in the state file, which means this will get ugly..

**Comment by Zombeaver** [comment]
That's unfortunate :(

**Comment by Zombeaver** [comment]
I'm not entirely certain that it even loads with the correct disk currently as you say.

I just tried Hitchhiker's Guide to the Galaxy ([csdb link](https://csdb.dk/release/?id=35816)) - it's a case like I described where you boot from the first disk then swap prior to starting the game. I started, swapped to the second disk, saved the state after the load was done, quit. Changed it to .state.auto but _didn't_ change the .m3u at all. When I start it back up it starts but as soon as you type anything it breaks. If I change the .m3u to Disk2.d64 > Disk1.d64, however, it works correctly.

**Comment by sonninnos** [comment]
Works fine here. Tried also the auto state route. Is the core version current?

But btw since the state does save the disk label and the Disk Control interface currently parses that even though it does not use it for anything, it looks like the frontend index sync might be simple to pull off after all. I'll try looking at it soonish.

**Comment by Zombeaver** [comment]
It is the current core, yes. I don't know what to tell you. I did what I described and that's the behavior I'm getting (and have always gotten).

Basically the disk that needs to be inserted at time of state save/load needs to be the first disk in the .m3u list otherwise it doesn't work. It's loading the order that's listed in the .m3u regardless of what's inserted at time of state save. It *loads* the state, it doesn't just crash immediately, but it clearly doesn't have the right disk inserted because the game just breaks when you try to do anything afterwards. Meanwhile swapping them around in the .m3u fixes it. I can record a video if that would be helpful.

**Comment by sonninnos** [comment]
Yes I do understand, but it simply does not happen here. The only way I can make it happen is if I explicitly eject and reinsert the disk manually after the state is loaded.

**Comment by Zombeaver** [comment]
Could it make any difference that I'm loading from command-line? I don't know why it would but I'm just throwing stuff out there.

I have everything to setup to run from bats with things like:

..\..\Retroarch\Retroarch.exe -L "..\..\Retroarch\cores\vice_x64_libretro.dll" "Hitchhiker's Guide to the Galaxy.m3u"

**Comment by sonninnos** [comment]
Nah, I also load from the command line.

And the m3u is simply:
```
Hitchhiker's Guide to the Galaxy-A.d64
Hitchhiker's Guide to the Galaxy-B.d64
```

Also I'm using x64sc, maybe that is the reason..

**Comment by Zombeaver** [comment]
I've renamed them to simply Disk1.d64 and Disk2.d64, but otherwise yes.

**Comment by sonninnos** [comment]
I disabled savestate file compression from RA and looked at the state file, and it clearly has the second disk label there "INFOCOM-S2".

Heh, also found an interesting message: "Side 2 being loaded into expansion ram. This will take about 10 minutes."

Currently there is no core option for memory expansions, but that can be enabled by adding `REU=1` to `vicerc`. The file goes to `system/vice/` and it also has the complete dump of all available options.

Like so for x64:
```
[C64]
REU=1
```

It sure did load for ages even with warp/fastforward, but looks like a worthy option. I'll add core options for enabling/resizing REU so that there is no need to hassle with `vicerc`.

**Comment by Zombeaver** [comment]
That's probably a good idea too, although I'm more concerned with why I seem to be experiencing different behavior than you.

Based on what I'm experiencing, something is not working as intended (for me). I don't really know what more information I can provide but if there's something you want me to test or do let me know. All I know is that it's not working correctly unless I manually adjust the .m3u after the fact to match what's inserted at time of the save.

**Comment by Zombeaver** [comment]
I am using a slightly older version of Retroarch itself, if you think that would make any difference. I believe it's 1.7.9.

**Comment by sonninnos** [comment]
I can't really think of anything else except also examining the state file and searching for the label. And the uncompressed state size should be roughly 2 * disk size (no idea why).

And sure that version can make a difference, since stuff changes all the time. I'm using the current nightly (1.9.0). But I'm not aware of any option that would enforce the DC index after state load.

**Comment by Zombeaver** [comment]
Okay, here's the state. 

[Hitchhiker's Guide to the Galaxy.state.auto.zip](https://github.com/libretro/vice-libretro/files/5431997/Hitchhiker.s.Guide.to.the.Galaxy.state.auto.zip)

For reference, here are some screenshots of what I'm seeing, specifically.

After restarting and loading from the state with an m3u with the original order (Disk1.d64 > Disk2.d64) and typing "get up":

![RetroArch Screenshot 2020 10 23 - 18 28 17 49](https://user-images.githubusercontent.com/24979172/97059097-915f0f00-155d-11eb-98a4-5a5b463b13ea.png)

After restarting and loading from the state with an m3u with a reversed disk order (Disk2.d64 > Disk1.d64) and typing "get up":

![RetroArch Screenshot 2020 10 23 - 18 30 13 10](https://user-images.githubusercontent.com/24979172/97059212-d2572380-155d-11eb-818f-6d2776a83fd0.png)

I'll try updating Retroarch just in case that make any difference here.

**Comment by sonninnos** [comment]
Huh, the state size is only around the memory size, as in it does not have the disk data included, so no wonder it won't insert it. Which means the core version simply cannot be current, since it explicitly forces the disk data saving when the drive type is 1541. It has done that for a long time. Check the version by git commit hash, not VICE version 3.3.

And yes I get the exact same garble data when I manually eject/reinsert after loading the state.

**Comment by Zombeaver** [comment]
It's only 7 days old. It's 254f5b4.

![image](https://user-images.githubusercontent.com/24979172/97060382-27e0ff80-1561-11eb-9015-d513b5cd22ac.png)

**Comment by sonninnos** [comment]
Uhh, this makes zero sense.. d64 files will set 1541 drive type, so that should be impossible. And if I can't replicate it, I can't fix it..

**Comment by Zombeaver** [comment]
Actually, wait a second...

When the game prompts for drive type of 1541 or 1571 I've been choosing 1571 because that _doesn't require TrueDrive to work_ (whereas the 1541 option does). When I turned on TrueDrive and chose 1541 and went through the process again, everything seems to be working as intended without swapping the .m3u around.

**Comment by sonninnos** [comment]
Ah, I never even tried that option since the disk image is d64. But TDE must be the reason then. I did try to disable that, but indeed it didn't load at all then, so I assumed it played no role..

**Comment by Zombeaver** [comment]
Okay so I tried again using the 1571 option but with TDE enabled again and it worked there too. So it's definitely the result of TDE being disabled.

Is there some way to... make it temporarily toggle TDE as part of the state save process so that it dumps what it needs? I'm assuming not since it's probably too late at that point.

This is all for my massive ongoing project of [C64 Dreams](https://forums.launchbox-app.com/topic/49324-c64-dreams-massive-curated-c64-collection/). I have a spreadsheet with details for every game in the set and as of my current WIP build only 125 (5.76%) out of 2169 games in the collection actually require TrueDrive - it's something that I'm always going to have disabled unless it's actually necessary. If there's no way around this issue if TDE is disabled, I guess I'll just have to stick with my current method of swapping .m3us around as needed.

**Comment by sonninnos** [comment]
Good news! I managed to trick the state saving by silently forcing TDE on during save and load, and then restoring the actual setting. Seems to be working, but more testing is required.

**Comment by Zombeaver** [comment]
> Good news! I managed to trick the state saving by silently forcing TDE on during save and load, and then restoring the actual setting. Seems to be working, but more testing is required.

Ha! It actually _does_ work! Kinda shocked it was as simple as that.

It definitely kept on trucking as expected this go round with TDE off in the options. I'll have to mess around with some more.

Is this likely to break existing states? I have quite a few auto-loading states setup to bypass some annoying or fiddly bits.

**Comment by Zombeaver** [comment]
So far so good on everything I've tested, both loading old states and creating new ones.

**Comment by sonninnos** [comment]
Excellent, and here is more good news! This version should now switch the DC index accordingly. Of course as long as the disks have distinct labels. At least works with the Hitchhiker. Please test if can you find problems with it before PR action.

[vice_x64_libretro.zip](https://github.com/libretro/vice-libretro/files/5433088/vice_x64_libretro.zip)

**Comment by Zombeaver** [comment]
Hmm... maybe I'm missing something or misunderstanding, but it doesn't seem like it's working. If I eject the disk it says it's on disk 1 (saved while on disk 2). That's what it shows in the core's status bar anyway. I'm not sure about RA's disk status messages because I would expect it to say what I'm currently seeing whether it's working correctly or not (next disk hotkey says disk 2 and previous says disk 1, but it would do that anyway whether it's already on 2 or not, because it doesn't circle around). I'll test with a 3+ disk game and see what it says.

![image](https://user-images.githubusercontent.com/24979172/97090959-359c9080-1606-11eb-82a9-8be2195a927c.png)

**Comment by Zombeaver** [comment]
Actually, I take it back - it does say Disk 2 in the status bar, but you have to eject and then reinsert (the same disk) for it to actually update and say that it's on 2 (which is fairly unintuitive).

**Comment by Zombeaver** [comment]
So this is a little odd - I tried out a 5 disk game, I saved while Disk 3 was inserted. Quit and reloaded and the status bar says Disk 1 when I eject, then I press next disk and it says Disk 5 is inserted. I would expect it to say Disk 4, I'm not sure why it went to 5. If I load the same state, eject, and press previous disk it says Disk 4 is inserted. So something isn't quite right there.

**Comment by sonninnos** [comment]
Well, like I said, it all depends on the disk labels. Try looking at the disks with DirMaster or with the regular "$" route, and if the label is the same, there is nothing that can be done, I reckon. What is the game and version in question?

Ejecting will show the wrong disk no matter what, but insert without changing index will not. 

The menu index is on the same page after state load:
![retroarch_2020_10_24_22_14_23_866](https://user-images.githubusercontent.com/45124675/97091693-5dabe400-1646-11eb-85ab-b1023ad7ccf3.png)

**Comment by Zombeaver** [comment]
Tested with a different 5 disk game, saved with disks 2 and 3 inserted and it's loading disk 1 on both, so I'm not sure what's going on.

It should probably be stated since it may be relevant that in both this scenario ([Portal](http://www.gamebase64.com/game.php?id=5836&d=18&h=0)) and the previous one ([Dragonworld](http://www.gamebase64.com/game.php?id=2329&d=18&h=0)) the game hadn't actually loaded any data from from disks 2-5, I just tested by swapping to those disks after the game had already loaded and then saving states. It's behaving differently between the two though anyway though, so something's odd.

**Comment by sonninnos** [comment]
Ok, well Portal disk labels are all just a bunch of garbled mess, so no wonder there is no match.

And Dragon World has all disks with the same label "ASS PRESENTS", so it also can not work..

**Comment by Zombeaver** [comment]
This is what it shows for Portal:

![image](https://user-images.githubusercontent.com/24979172/97091746-2f111780-160c-11eb-96b4-426b5e9392d2.png)

This is what it shows for Dragonworld:

![image](https://user-images.githubusercontent.com/24979172/97091790-726b8600-160c-11eb-8c00-3e51609206c5.png)

**Comment by Zombeaver** [comment]
Yeah, let me try renaming the labels and see what happens.

**Comment by sonninnos** [comment]
Aye, looks like the only fool proof method would be to edit the savestate file format to add the filename somewhere, and that is very likely a can of worms..

**Comment by Zombeaver** [comment]
Yep, sure enough. I renamed the labels to Dragonworld A - E and it worked as expected. UGH. One more thing that needs doing haha. At least it's working though. I guess my only concern is whether or not it's safe to do that across the board. I'm not sure if that's something that's specifically checked by any games when loading.

I don't actually have the statusbar enabled by default so having the wrong disk label before making any changes isn't actually the end of the world for me (because all you'll see is RA's disk index message), but I could see that being confusing if someone did use it. I don't know if there's a way around that but like I said, I guess not that huge of a deal.

**Comment by Zombeaver** [comment]
And I guess the next question is, is the only thing that matters that the labels be unique? In other words, it doesn't have to be in some sort of alphabetical or numerical order or anything like that just so long as none of them are duplicates?

None of the ones for Portal are duplicates, is the reason I'm asking. They are a bunch of gibberish but none of them are the same gibberish.

**Comment by sonninnos** [comment]
Ok, the index sync will update the statusbar label now too.

The labels are converted to readable ASCII, so gibberish pretty much means they are empty, as in the same. I guess I could remove that sanitizing now and keep the actual chars for these matching purposes, since the labels are not used in the widget messages anymore, because the frontend shows the disk filenames on index change, and showing both messages is problematic.

**Comment by Zombeaver** [comment]
> Aye, looks like the only fool proof method would be to edit the savestate file format to add the filename somewhere, and that is very likely a can of worms..

If this is actually possible I think that would actually be preferable, just so that there's no risk of potentially breaking anything by altering disk labels and ensures that it "just works" out of the box. I'm crazy enough to actually put in the work to go through and change them all if need be but most people aren't.

**Comment by sonninnos** [comment]
Yep, and including the filename in the state would require a lot more work to keep everything in order and compatible, and the payoff is pretty minimal, as in purely visual now after the TDE revelation..

**Comment by Zombeaver** [comment]
I wouldn't really consider that minimal to be honest. I've been going through my multi-disk stuff and it's not exactly uncommon to see the same label between multiple disks. Seems like it's at least 50/50.

And with some further testing it looks like some games definitely _do_ look at the label when attempting to load. I tried changing the labels on Hitchhiker's Guide to the Galaxy to be "HGTTG" on both disks to see what would happen and it wouldn't even load the second disk afterwards.

Making it include the filename seems like the only safe way to ensure that they're indexed correctly, and will work across the board.

**Comment by sonninnos** [comment]
Oh crap, the index change should have only changed the visual index and not the disk, but now it also changes the disk, so sure, that is a clear problem.

**Comment by Zombeaver** [comment]
You lost me. Why wouldn't it change the disk? Is that the whole point? If I load a state that's saved while disk 3 out of 5 is inserted, I would expect that when I load that state that disk 3 is then the current inserted disk.

**Comment by sonninnos** [comment]
Yes, the point of this exercise is to make the frontend sync up with what the state dictates. Without these latest additions the software would work fine, but the user would have to "reset" the situation on disk changes. As in next and previous are whatever unless started from the beginning.

Even VICE itself won't sync up properly with the savestates, so I had to jump through hoops to get this far.

Looks like all I can do then is to make the match as strict as possible so that it won't break anything, as in do nothing if there are multiple matches, or just scrap the whole thing and leave it at that TDE addition, which is good enough and better than nothing I suppose..

**Comment by Zombeaver** [comment]
The ideal behavior would be:

I have a 5 disk game. I can be on any disk and save a state - we'll say 3 - and when I later load that state it's loaded with disk 3 inserted and pressing next disk switches to (and displays) disk 4, and previous disk switches to (and displays) disk 2, regardless of what's in the .m3u.

The only way I know of to actually make that happen safely would be to make it index based on the actual filenames, not the disk label. Regardless of the hassle involved in going through and changing them where needed, in only a brief amount of testing I've seen that that's not actually universally safe to do - some games definitely do check for a specific label when loading, so that's not really an option because you'll end up breaking things in the process.

**Comment by Zombeaver** [comment]
If for some reason that isn't possible (I'm really hoping that it is), then I think you'd probably want to just revert everything except for the TDE change, because you're likely to break more stuff in the long run if it's completely dependent on the label - based on examining the multi-disk games in my collection it's fairly common to see the same label used between multiple disks. I think it would be better to have it always display 1 at the start than it would be to (occasionally) display a different (wrong) disk - at least that way it's consistent.

**Comment by sonninnos** [comment]
Phew, luckily I don't like giving up easily, heh.

Now new saves will not save the disk data, but only the filename instead. Which means state sizes with TDE enabled will be on par with the size before the TDE force hack. I tried my best to make sure old saves are not affected.

[vice_x64_libretro.zip](https://github.com/libretro/vice-libretro/files/5434140/vice_x64_libretro.zip)

**Comment by Zombeaver** [comment]
Awesome! I'll try it out. Thanks!

**Comment by Zombeaver** [comment]
I think you've got something here! So far so good with everything I've tested - all seem to be working exactly as I'd hoped, including the previously problematic cases, and the index is displaying correctly from the start. You're the man! More testing needed but looks very good so far.

The only thing I've encountered so far that might need some tweaking (and I think this has probably always been like this, and not directly related to any of these changes) is that when you press reset (End) it seems to keep the same disk inserted that's in at the time of reset. I suppose that's to be expected but it might be nice to make it (perhaps optionally?) revert to the first disk in the .m3u when triggering a reset. This is by no means a huge deal - it will still work correctly on the initial start regardless because I won't be making any adjustments to the .m3us anymore, but it would be nice to have the option of a "fresh start" with just a single button press.

**Comment by Zombeaver** [comment]
I did encounter something strange while testing - one game that I knew I had working ([The Guild of Thieves](https://csdb.dk/release/?id=56738)) no longer seems to boot. It requires TrueDrive, which I have on, but it doesn't start. I went back and tested an old version of the core that I happened to have backed up and it worked. Swapped to the newest one (and the current one in the updater too, so it's not from these changes) and it doesn't work.

With the current one I'm getting a "DOS Error: 31 Syntax Error, 00, 00" error and with the old one it's working.

[vice_x64_libretro-old.zip](https://github.com/libretro/vice-libretro/files/5434256/vice_x64_libretro-old.zip)

**Comment by sonninnos** [comment]
> when you press reset (End) it seems to keep the same disk inserted that's in at the time of reset. I suppose that's to be expected but it might be nice to make it (perhaps optionally?) revert to the first disk in the .m3u when triggering a reset. 

Yes that is by design, meant primarily for non-disk cases. Such that it is possible to make playlists of PRGs and CRTs too. Those simply would not work if it always resets to the first index. Also there are some ZIPs in gamebase64 that have multiple versions of the same game, and this allows simple access to them without manual file hassle. No-Intro might have some ZIPs too that does not play well with alphabetical sorting.

But I could make the Quick Menu -> Restart reset the index and leave the core reset hotkey as is, since the main restart should absolutely do that anyway IMO.


> With the current one I'm getting a "DOS Error: 31 Syntax Error, 00, 00" error and with the old one it's working.

Damn, the working version must be very old then. I managed to replicate the same error even with standalone VICE, and luckily found the reason at the first try. It is caused by the option "Virtual device traps" which is forcefully enabled currently, and disabling won't work even with `vicerc`. I enabled that a long time ago, because the printer device would not work otherwise (it outputs to `vice_printer.txt` in `saves`). Looks like I have to make it optional after all.

The core used to handle it in a way that enabling TDE would disable the virtual traps and disabling TDE would enable them. Now we finally know that there was a reason for that behavior. Thanks for spotting!

---

Only thing I've found so far with this savestate hack is that enabling rewind starts flashing the statusbar drive LED, so I have to try to sort that first and close this issue before moving to the next ones.

**Comment by Zombeaver** [comment]
> Yes that is by design, meant primarily for non-disk cases. Such that it is possible to make playlists of PRGs and CRTs too. Those simply would not work if it always resets to the first index. Also there are some ZIPs in gamebase64 that have multiple versions of the same game, and this allows simple access to them without manual file hassle. No-Intro might have some ZIPs too that does not play well with alphabetical sorting.

Ah okay, that's a good point. Fair enough.

> But I could make the Quick Menu -> Restart reset the index and leave the core reset hotkey as is, since the main restart should absolutely do that anyway IMO.

Yep, that should work.

> It is caused by the option "Virtual device traps" which is forcefully enabled currently, and disabling won't work even with vicerc. I enabled that a long time ago, because the printer device would not work otherwise (it outputs to vice_printer.txt in saves). Looks like I have to make it optional after all.

Interesting! Well, I'm glad you were able to pin it down so quickly!

> Only thing I've found so far with this savestate hack is that enabling rewind starts flashing the statusbar drive LED

Odd...

> so I have to try to sort that first and close this issue before moving to the next ones.

Yeah, sorry didn't mean to throw more stuff onto the pile, was just stuff that came to mind or that I came across while testing. Thank you _very_ much for the work you're doing on this!

**Comment by sonninnos** [comment]
Cheers and no worries, I like challenges and I like Commodore, heh.

Unfortunately the statusbar drive track number flashing while rewind is enabled is only a visual warning that something really is wrong, which is that drive access will be toast. Naturally in normal cases no one would save during disk access, but rewind saves constantly and it has to work too. Of course it just can't be that easy..

**Comment by Zombeaver** [comment]
Seems to be working although it has broken some of my older states, including some that were saved before any of these changes. That's kindof unfortunate but not the end of the world I guess. I'll have to go through all my games that use them and double-check.

**Comment by sonninnos** [comment]
Crap, that is unacceptable. My saves have worked so far, but I'll keep testing.

**Comment by Zombeaver** [comment]
Do you want me to send you the ones that crash? I don't know if that would help or not. It's not a ton of them, but there have been a few so far.

**Comment by sonninnos** [comment]
I went through all of my d64 based states, and had zero issues. Tried also disabling TDE, and even some multiloaders loaded fine. Maybe this is because they are saved with TDE enabled or something to do with x64 vs x64sc..

Sure, send one if it is something that is certified to crash and not some random thing.

**Comment by Zombeaver** [comment]
Here are what I've run into so far. This is the state + disks. These all insta-crash on boot.

[Crashing States.zip](https://github.com/libretro/vice-libretro/files/5436003/Crashing.States.zip)

> The most important thing is of course that none of the new saves crash.

No, I haven't run into any crashing there thus far.

**Comment by sonninnos** [comment]
Whoops, these changes totally should not have any effect on anything else excepts D64s, so that was a goof which was only revealed with x64, because I did try a few carts too. Cabal and Castle of Terror states do not crash anymore, neither does Genghis Khan but it simply refuses to work, no idea why.. But if there is no more similar surprises maybe we can finally seal the deal with this version:
[vice_x64_libretro.zip](https://github.com/libretro/vice-libretro/files/5436056/vice_x64_libretro.zip)

**Comment by Zombeaver** [comment]
> neither does Genghis Khan but it simply refuses to work

Sorry, I should have mention, that one requires NTSC. It says in the cracktro that it's PAL fixed but it's not really, as far as I can tell. It works fine with C64 NTSC though.

I found some more (attached below). I'll retest with the new version.

Regarding Dragon's Lair I and II (which is a single easyflash crt) in that attachment, I use .cmds to start everything, and I have separate ones for Dragon's Lair and Dragon's Lair II which is why there are separate states for that same cart - they're loaded after the selection of I or II has already been made so that it starts right on those games rather than prompting to choose one.

[Crashing States 2.zip](https://github.com/libretro/vice-libretro/files/5436071/Crashing.States.2.zip)

So far I've gone through about 40% of the collection and the ones that have been attached so far are the only ones that crashed.

**Comment by Zombeaver** [comment]
Yep, sure enough none of those crash in the newest version. A very good sign!

**Comment by sonninnos** [comment]
I just meant the autostate not working, the game itself did launch in PAL. Might be another x64 vs x64sc thing. Why even use the fast version? The accurate one is not that much slower with current hardrware, and VICE 3.4 skips building the fast one by default already.

**Comment by Zombeaver** [comment]
> the game itself did launch in PAL

It does, but it doesn't get passed the trainer screen.

> Why even use the fast version? The accurate one is not that much slower with current hardrware, and VICE 3.4 skips building the fast one by default already.

Well keep in mind that not everyone has great hardware and I've had to tone down a few default settings for people even with the fast core. I used to have a default setting of Hard GPU Sync: On / Frame Delay 3 for example but I've talked to people that have performance problems even at Frame Delay 1 (so the default is 0 now).

I'm not _opposed_ to switching to sc at all (and if it were only me, I would have already), but I'm trying to make it usable for as many people as I can. I've considered switching it over already though, and may still before the next release. Having tested 2000+ games on the fast core though, I can say that _it does work_. I've never run into a situation of a game that just refused to work (with the right combination of TDE, region, etc.) that spontaneously started working with sc.

**Comment by sonninnos** [comment]
Hmm, I got very far past the cracktro with both x64 and x64sc.

Sure there are plenty of low-power machines still out there, and luckily after the zoom crop replacement I did to this core there are no longer issues with some sprites disappearing near the VIC-II borders like with VICE internal border settings in x64. And indeed AFAIK the accuracy stuff only concerns VIC-II stuff. And fast sure is a wee bit faster when warping/fastforwarding.

**Comment by Zombeaver** [comment]
> Hmm, I got very far past the cracktro with both x64 and x64sc.

Not the cracktro, the trainers. There's a _ridiculously long_ intro to the game, then some trainers, then the title screen, then the game starts. It's what the state is there for to begin with - I have separate .cmds that people can use to start the game, either with or without the intro (in which case it loads the state to the trainer screen). With it set to PAL it will get through the entire intro, get to the trainers, then you'll get to the title screen, then just a black screen. With NTSC it works normally. I just tested again and it's because of TDE being off.

> And fast sure is a wee bit faster when warping/fastforwarding.

That's about 3-4x faster for me.

That adds up _real quick_ when you're testing stuff haha

**Comment by sonninnos** [comment]
Ah, so it seems. Super lengthy unskippable intro sure is not reasonable. Especially when RUN/STOP crashes it. Luckily it can be skipped with `LOAD"K6",8,1` and indeed black screen awaits in PAL even with TDE on, so that direct route skips the PAL fix then I guess.

I'll be adding a few more small things to the same PR, so there is still some time to find problems in the savestate hack.

**Comment by Zombeaver** [comment]
Okay, sounds good. I'll continue testing and let you know if I run into anything, but so far so good.
